src/mca/ptl/tcp/ptl_tcp_component.c

```diff
@@ -1896,16 +1845,78 @@ static void process_cbfunc(int sd, short args, void *cbdata)
     /* validate the connection */
     cred.bytes = pnd->cred;
     cred.size = pnd->len;
-    PMIX_PSEC_VALIDATE_CONNECTION(rc, peer, NULL, 0, NULL, NULL, &cred);
-    if (PMIX_SUCCESS != rc) {
-        pmix_output_verbose(2, pmix_ptl_base_framework.framework_output,
-                            "validation of tool credentials failed: %s",
-                            PMIx_Error_string(rc));
-        PMIX_RELEASE(peer);
-        pmix_list_remove_item(&pmix_server_globals.nspaces, &nptr->super);
-        PMIX_RELEASE(nptr);  // will release the info object
+    PMIX_PSEC_VALIDATE_CONNECTION(reply, peer, NULL, 0, NULL, NULL, &cred);
+    if ( reply == PMIX_ERR_READY_FOR_HANDSHAKE ) {
+        u32 = htonl(reply);
+        rc = pmix_ptl_base_send_blocking(pnd->sd, (char*)&u32, sizeof(uint32_t));
+        if (PMIX_SUCCESS != rc ) {
+            PMIX_ERROR_LOG(rc);
+            info->proc_cnt--;
+            pmix_pointer_array_set_item(&pmix_server_globals.clients, peer->index, NULL);
+            PMIX_RELEASE(peer);
+            CLOSE_THE_SOCKET(pnd->sd);
+            PMIX_RELEASE(pnd);
+            return;
+        }
+        PMIX_PSEC_SERVER_HANDSHAKE(reply, peer, NULL, 0, NULL, NULL, &cred);
+    }
+    pmix_output_verbose(2, pmix_ptl_base_framework.framework_output,
+                        "client connection validated");
+
+    /* send the status so they don't hang */
+    u32 = ntohl(reply);
+    if (PMIX_SUCCESS != (rc = pmix_ptl_base_send_blocking(pnd->sd, (char*)&u32, sizeof(uint32_t)))) {
+        PMIX_ERROR_LOG(rc);
         CLOSE_THE_SOCKET(pnd->sd);
-        goto done;
+        PMIX_RELEASE(pnd->peer);
+        PMIX_RELEASE(pnd);
+        PMIX_RELEASE(cd);
+        return;
+    }
+
+    /* if we got an identifier, send it back to the tool */
+    if (pnd->need_id) {
```

Comment:
Also, here
https://github.com/pmix/pmix/pull/1240/files#diff-2f3e46bf48df10ec8101693fba191ef3R1283
I don't see that this case is expected in any way.
Only 2 receives are performed instead of 4.

Reply:
Again, I'd have to look at the broader context to understand the scenario. Will do so upon return from vacation.